### PR TITLE
IGNITE-21513 Fixed ActiveActorTest#testChangeLeaderForce is flaky

### PR DESCRIPTION
**Pull request by denis-chudov**
https://issues.apache.org/jira/browse/IGNITE-21513

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

**Review by vldpyatkov — APPROVED** [review]

**Review by alievmirza — COMMENTED** [review]
You have a lot of "yellow" code, meaning that idea static analyser reports a lot of warns. Please take a look on them.

Could you please also describe, what exactly was changed in the test, so it stop being flaky?

**Comment by denis-chudov** [comment]
> You have a lot of "yellow" code, meaning that idea static analyser reports a lot of warns. Please take a look on them.
> 
> Could you please also describe, what exactly was changed in the test, so it stop being flaky?

This information was in comment to [IGNITE-21381](https://issues.apache.org/jira/browse/IGNITE-21381). Added this to ticket description.

File: modules/placement-driver/src/integrationTest/java/org/apache/ignite/internal/placementdriver/ActiveActorTest.java [hunk 1]
```diff
@@ -99,53 +62,43 @@
...
-    private static final int PORT_BASE = 1234;
-
-    @InjectConfiguration
-    protected RaftConfiguration raftConfiguration;
-
-    /** RPC executor. */
-    protected ScheduledExecutorService executor = new ScheduledThreadPoolExecutor(20, new NamedThreadFactory("Raft-Group-Client", log));
-
+public class ActiveActorTest extends AbstractTopologyAwareGroupServiceTest {
     private Map<String, PlacementDriverManager> placementDriverManagers = new HashMap<>();
```
Comment (alievmirza) [inline modules/placement-driver/src/integrationTest/java/org/apache/ignite/internal/placementdriver/ActiveActorTest.java:68]: please make this field final

File: modules/replicator/src/test/java/org/apache/ignite/internal/raft/client/TopologyAwareRaftGroupServiceTest.java [hunk 2]
```diff
@@ -17,433 +17,33 @@
...
-                peersAndLearners(clusterServices, isServerAddress, nodes),
-                true,
-                executor,
-                new LogicalTopologyServiceTestImpl(localClusterService),
-                eventsClientListener,
-                notifyOnSubscription,
-                commandsMarshaller
-        ).join();
+    @Override
+    protected void afterClusterInit(String leaderName) throws InterruptedException {
```
Comment (alievmirza) [inline modules/replicator/src/test/java/org/apache/ignite/internal/raft/client/TopologyAwareRaftGroupServiceTest.java:36]: please remove this redundant `InterruptedException` here and below

File: modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+    /**
+     * Stops cluster.
+     *
+     * @param clusterServices Cluster services.
+     * @param raftServers     RAFT services.
+     * @param raftClients     RAFT clients.
+     * @param nodes           Node count.
+     * @throws Exception If failed.
+     */
+    private void stopCluster(
```
Comment (alievmirza) [inline modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java:340]: please make it static

File: modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,522 @@
...
+            HashMap<NetworkAddress, ClusterService> clusterServices,
+            HashMap<NetworkAddress, JraftServerImpl> raftServers,
+            List<TopologyAwareRaftGroupService> raftClients,
+            int nodes
+    ) throws Exception {
+        if (raftClients != null) {
+            raftClients.forEach(client -> client.shutdown());
```
Comment (alievmirza) [inline modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java:347]:
```suggestion
            raftClients.forEach(TopologyAwareRaftGroupService::shutdown);
```

File: modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,522 @@
...
+        CompletableFuture<Void> future = client.subscribeLeader((node, term) -> {
+            callback.onLeaderElected(node, term);
+
+            log.info(logMessage, node);
+        });
+
+        assertThat(future, willCompleteSuccessfully());
+    }
+
+    private void unsubscribeLeader(TopologyAwareRaftGroupService client) {
```
Comment (alievmirza) [inline modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java:517]: may be static

File: modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,522 @@
...
+                                toSet()));
+    }
+
+    /**
+     * Generates a node address for each node.
+     *
+     * @param nodes Node count.
+     * @return List on network addresses.
+     */
+    private static List<NetworkAddress> getNetworkAddresses(int nodes) {
```
Comment (alievmirza) [inline modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java:500]: you could use `org.apache.ignite.internal.network.utils.ClusterServiceTestUtils#findLocalAddresses`, up to you

File: modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,522 @@
...
+
+        assertEquals(raftClientNoInitialNotify.leader().consistentId(), leaderRef.get().name());
+
+        stopCluster(clusterServices, raftServers, List.of(raftClient, raftClientNoInitialNotify), nodes);
+
+        clientClusterService.stop();
+    }
+
+    @Test
+    public void testChangeLeaderForce(TestInfo testInfo) throws Exception {
```
Comment (alievmirza) [inline modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java:248]: a lot of code here and in `testChangeLeaderWhenActualLeft` is duplicated, like 50 lines in the beginning, please refactor

File: modules/placement-driver/src/integrationTest/java/org/apache/ignite/internal/placementdriver/ActiveActorTest.java [hunk 8]
```diff
@@ -193,453 +144,22 @@ private boolean checkSingleActiveActor(String leaderName) {
...
-        @Override
-        public boolean onSnapshotLoad(Path path) {
-            return true;
-        }
+        when(entry.empty()).thenReturn(true);
 
-        @Override
-        public void onShutdown() {
-        }
+        return entry;
```
Comment (alievmirza) [inline modules/placement-driver/src/integrationTest/java/org/apache/ignite/internal/placementdriver/ActiveActorTest.java:162]: Below this code we have `LogicalTopologyServiceTestImpl` and also we have exactly the same `org.apache.ignite.internal.topology.LogicalTopologyServiceTestImpl`. Please do something with that

File: modules/placement-driver/src/integrationTest/java/org/apache/ignite/internal/placementdriver/ActiveActorTest.java [hunk 9]
```diff
@@ -99,53 +62,43 @@
...
-
+public class ActiveActorTest extends AbstractTopologyAwareGroupServiceTest {
     private Map<String, PlacementDriverManager> placementDriverManagers = new HashMap<>();
 
     @Mock
     MetaStorageManager msm;
 
+    @BeforeEach
+    public void setUp() {
+        when(msm.recoveryFinishedFuture()).thenReturn(completedFuture(0L));
```
Comment (alievmirza) [inline modules/placement-driver/src/integrationTest/java/org/apache/ignite/internal/placementdriver/ActiveActorTest.java:75]: please check if it is correct to return 0 here

File: modules/placement-driver/src/integrationTest/java/org/apache/ignite/internal/placementdriver/ActiveActorTest.java [hunk 10]
```diff
@@ -193,453 +144,22 @@ private boolean checkSingleActiveActor(String leaderName) {
...
-
-        raftClient.refreshLeader().get();
-
-        assertEquals(raftClient.leader().consistentId(), leaderRef.get().name());
-
-        stopCluster(clusterServices, raftServers, List.of(raftClient, raftClientNoInitialNotify), nodes);
-        clientClusterService.stop();
+    @Override
+    protected void afterClusterInit(String leaderName) throws InterruptedException {
+        assertTrue(waitForCondition(() -> checkSingleActiveActor(leaderName), 10_000));
```
Comment (alievmirza) [inline modules/placement-driver/src/integrationTest/java/org/apache/ignite/internal/placementdriver/ActiveActorTest.java:149]: please extract `10_000` to a common timeout value all over your changes

File: modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java (hunk 7 above)
Comment (alievmirza) [reply modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java:248]: also please add some comments or make code more readable, it is really hard to understand what is going on in these tests

File: modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,522 @@
...
+            }
+
+            if (addr.port() == clientPort) {
+                assertTrue(isServerAddress.test(addr));
+
+                raftClient = startTopologyAwareClient(cluster, clusterServices, isServerAddress, nodes, eventsClientListener, true);
+            }
+        }
+
+        return raftClient;
```
Comment (alievmirza) [inline modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java:441]: `raftClient` could be null, is it ok?

File: modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,522 @@
...
+
+        return TopologyAwareRaftGroupService.start(
+                GROUP_ID,
+                localClusterService,
+                FACTORY,
+                raftConfiguration,
+                peersAndLearners(clusterServices, isServerAddress, nodes),
+                true,
+                executor,
+                new LogicalTopologyServiceTestImpl(localClusterService),
```
Comment (alievmirza) [inline modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java:476]: I cannot understand why we create `new LogicalTopologyServiceTestImpl(localClusterService)` here and in `org.apache.ignite.internal.placementdriver.ActiveActorTest#afterNodeStart`? Shouldn't be here only one `LogicalTopologyService`?

**Review by alievmirza — CHANGES_REQUESTED** [review]

File: modules/placement-driver/src/integrationTest/java/org/apache/ignite/internal/placementdriver/ActiveActorTest.java (hunk 9 above)
Comment (denis-chudov) [reply modules/placement-driver/src/integrationTest/java/org/apache/ignite/internal/placementdriver/ActiveActorTest.java:75]: it doesn't matter, it doesn't start the node and perform the actual recovery

File: modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java (hunk 11 above)
Comment (denis-chudov) [reply modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java:441]: added assertions

File: modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java (hunk 7 above)
Comment (denis-chudov) [reply modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java:248]: moved the cluster initialization to the separate method

File: modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java (hunk 12 above)
Comment (denis-chudov) [reply modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java]: fixed

File: modules/placement-driver/src/integrationTest/java/org/apache/ignite/internal/placementdriver/ActiveActorTest.java (hunk 10 above)
Comment (denis-chudov) [reply modules/placement-driver/src/integrationTest/java/org/apache/ignite/internal/placementdriver/ActiveActorTest.java]: done

**Review by alievmirza — APPROVED** [review]

**Pull request merged**